akka-remote/src/main/scala/akka/remote/artery/ArteryTcpTransport.scala

```diff
@@ -0,0 +1,365 @@
+        //       related to give-up-system-message-after ?
+        RestartFlow.withBackoff[ByteString, ByteString](1.second, 5.seconds, 0.1)(flowFactory)
+      } else {
+        // Best effort retry a few times
+        RestartFlow.withBackoff[ByteString, ByteString](1.second, 5.seconds, 0.1, maxRestarts = 3)(flowFactory)
+      }
+    }
+
+    Flow[EnvelopeBuffer]
+      .map { env ⇒
+        val bytes = ByteString(env.byteBuffer)
+        bufferPool.release(env)
+        bytes
+      }
+      .recoverWithRetries(1, { case ArteryTransport.ShutdownSignal ⇒ Source.empty })
+      .via(connectionFlowWithRestart)
+      .map(_ ⇒ throw new IllegalStateException(s"Unexpected incoming bytes in outbound connection to [${outboundContext.remoteAddress}]"))
+      .toMat(Sink.ignore)(Keep.right)
+
+    // FIXME The mat value Future is currently never completed, because RestartFlow will retry forever, should it give up?
+    //       related to give-up-system-message-after ?
+
+  }
+
+  override protected def runInboundStreams(): Unit = {
+    implicit val mat = materializer
+    implicit val sys = system
+
+    val (controlStream, controlStreamCompleted) = runInboundControlStream()
+    val (ordinaryMessagesStream, ordinaryMessagesStreamCompleted) = runInboundOrdinaryMessagesStream()
+    val (largeMessagesStream, largeMessagesStreamCompleted) = {
+      if (largeMessageChannelEnabled)
+        runInboundLargeMessagesStream()
+      else
+        (
+          Flow[EnvelopeBuffer]
+          .map(_ ⇒ log.warning("Dropping large message, missing large-message-destinations configuration."))
+          .to(Sink.ignore),
+          Promise[Done]().future) // never completed, not enabled
+    }
+
+    val inboundStream: Sink[EnvelopeBuffer, NotUsed] =
+      Sink.fromGraph(GraphDSL.create() { implicit b ⇒
+        import GraphDSL.Implicits._
+        val partition = b.add(Partition[EnvelopeBuffer](3, env ⇒ {
+          env.byteBuffer.get(EnvelopeBuffer.StreamIdOffset).toInt match {
+            case `ordinaryStreamId` ⇒ 1
+            case `controlStreamId`  ⇒ 0
+            case `largeStreamId`    ⇒ 2
+          }
+        }))
+        partition.out(0) ~> controlStream
+        partition.out(1) ~> ordinaryMessagesStream
+        partition.out(2) ~> largeMessagesStream
+        SinkShape(partition.in)
+      })
+
+    val maxFrameSize =
+      if (largeMessageChannelEnabled) settings.Advanced.MaximumLargeFrameSize
+      else settings.Advanced.MaximumFrameSize
+
+    // FIXME error handling in case something fails in this part
+
+    val inboundConnectionFlow = Flow[ByteString]
+      .via(inboundConnectionsKillSwitch.flow)
+      .via(Framing.lengthField(fieldLength = 4, fieldOffset = EnvelopeBuffer.FrameLengthOffset,
+        maxFrameSize, byteOrder = ByteOrder.LITTLE_ENDIAN))
+      .map { frame ⇒
+        val buffer = ByteBuffer.wrap(frame.toArray)
```

Comment:
I guess this is the simplest way, or we could have it as a sealed trait with two impls and use ByteStrings. But I see quite a lot of places access the byte buffer

Reply:
yeah, some overhead of the copying, but I think it's "good enough"